### DISCUSSION TITLE
Animate mesh using absolute transformation matrix (without deep_copy)

### DISCUSSION THREAD

**Discussion by jakub-kaminski** [isl-org/Open3D#5116] [Q&A]
Hi,
I plan to animate several rigid bodies (a robot arm) represented as meshes.
I need to display each of them transformed using a absolute transformation (using DH table). It looks like I need to undo last Open3D mesh transformation a mesh every time after displaying the rigid body, so that the transformations do not accumulate. Is there a better solution with Open3D?

Bad ideas:
- Applying deep_copy of all the meshes every time robot is drawn. Issues: copying a lot of data and sending vertices to GPU every time rather than transformation matrix only.
- Keeping track on the previous absolute transformation to return object to its original loaded state. Operations: undo last transformation, apply new transformation and then redraw object. Sounds like potentially accumulating numerical errors & dirty workaround.

This use case might not be the best one for this library. I know that there are better tools to visualize robots.
Here I am looking for a replacement for Matlab that will enable robot visualization ("patch" objects on GPU with updated transformation matrix) and easy plotting.

Maybe I should look for another tool?

Thank you for your time and support!
Jakub

**Comment by RubensBenevides** [comment]
Hi jakub-kaminski. Some time ago I was dealing with the same problem, I used your second option. It's very slow, but it was the only way to animate my point clouds. In my master's degree I developed a model to correct the deviations that occur by sequential recording. In my dataset I have 900 clouds of a velodyne LiDAR collected by a moving robot, and I wanted to animate the correction, that is, from the poses with drift to poses without drift. To make the animation run at 30 FPS, with 10 seconds of animation, I needed to interpolate 300 poses (rigid 4x4 transformations) between 900 pairs of robot poses. The most complicated job of my life.

I implemented everything using numpy arrays, because the list is slow in python. With the 300x900 = 270000 poses ready to transform the 900 objects, I realized that every time the 900 clouds were transformed into a new frame of reference, I needed to transform everything back. The only way was to apply all the inverse transformations before applying the next 900 interpolated poses. The pose interpolation was done using quaternions. So I also needed to transform matrices into quaternions and quaternions into matrices.

Processes were so slow that I had to set code to take print screens out of preview every time the render refreshed. I use an i3-9400K (4.6GHz) + AMD-RX560. I will update my comment with the Youtube link when I post the videos.

**Reply by RubensBenevides** [reply]
Here is the video https://www.youtube.com/shorts/XAiHnahhgpg
